src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,244 @@
+begin
+  apply limit.hom_ext,
+  rintro ⟨⟨Y, f, hf⟩, ⟨Z, g, hg⟩⟩,
+  simp only [first_map, second_map, fork_map, limit.lift_π, limit.lift_π_assoc, assoc,
+    fan.mk_π_app, subtype.coe_mk, subtype.val_eq_coe],
+  rw [← P.map_comp, ← op_comp, pullback.condition],
+  simp,
+end
+
+def is_sheaf' (P : Cᵒᵖ ⥤ A) : Prop := ∀ (U : C) (R : presieve U) (hR : generate R ∈ J U),
+nonempty (is_limit (fork.of_ι _ (w R P)))
+
+/-- Impl. An auxiliary lemma to convert between sheaf conditions. -/
+def is_sheaf_for_is_sheaf_for' (P : Cᵒᵖ ⥤ A) (s : A ⥤ Type v)
+  [Π J, preserves_limits_of_shape (discrete J) s] (U : C) (R : presieve U) :
+  is_limit (s.map_cone (fork.of_ι _ (w R P))) ≃
+    is_limit (fork.of_ι _ (equalizer.presieve.w (P ⋙ s) R)) :=
+begin
+  apply equiv.trans (is_limit_map_cone_fork_equiv _ _) _,
+  apply (is_limit.postcompose_hom_equiv _ _).symm.trans (is_limit.equiv_iso_limit _),
+  { apply nat_iso.of_components _ _,
+    { rintro (_ | _),
+      { apply preserves_product.iso s },
+      { apply preserves_product.iso s } },
+    { rintro _ _ (_ | _),
+      { ext : 1,
+        dsimp [equalizer.presieve.first_map, first_map],
+        simp only [limit.lift_π, map_lift_pi_comparison, assoc, fan.mk_π_app, functor.map_comp],
+        erw pi_comparison_comp_π_assoc },
+      { ext : 1,
+        dsimp [equalizer.presieve.second_map, second_map],
+        simp only [limit.lift_π, map_lift_pi_comparison, assoc, fan.mk_π_app, functor.map_comp],
+        erw pi_comparison_comp_π_assoc },
+      { dsimp,
+        simp } } },
+  { refine fork.ext (iso.refl _) _,
+    dsimp [equalizer.fork_map, fork_map],
+    simp }
+end
+
+/-- The equalizer definition of a sheaf given by `is_sheaf'` is equivalent to `is_sheaf`. -/
+theorem is_sheaf_iff_is_sheaf' (P : Cᵒᵖ ⥤ A) :
+  is_sheaf J P ↔ is_sheaf' J P :=
+begin
+  split,
+  { intros h U R hR,
+    refine ⟨_⟩,
+    apply coyoneda_jointly_reflects_limits,
+    intro X,
+    have q : presieve.is_sheaf_for (P ⋙ coyoneda.obj X) _ := h X.unop _ hR,
+    rw ←presieve.is_sheaf_for_iff_generate at q,
+    rw equalizer.presieve.sheaf_condition at q,
+    replace q := classical.choice q,
+    apply (is_sheaf_for_is_sheaf_for' _ _ _ _).symm q },
+  { intros h U X S hS,
+    rw equalizer.presieve.sheaf_condition,
+    refine ⟨_⟩,
+    refine is_sheaf_for_is_sheaf_for' _ _ _ _ _,
+    apply is_limit_of_preserves,
+    apply classical.choice (h _ S _),
+    simpa }
+end
+
+/--
+For a concrete category `(A, s)` where the forgetful functor `s : A ⥤ Type v` preserves limits and
+reflects isomorphisms, and `A` has limits, an `A`-valued presheaf `P : Cᵒᵖ ⥤ A` is a sheaf iff its
+underlying `Type`-valued presheaf `P ⋙ s : Cᵒᵖ ⥤ Type` is a sheaf.
+
+Note this lemma applies for "algebraic" categories, eg groups, abelian groups and rings, but not
+for the category of topological spaces, topological rings, etc since reflecting isomorphisms doesn't
+hold.
+-/
+lemma sheaf_cond3 (P : Cᵒᵖ ⥤ A) (s : A ⥤ Type v)
```

Comment:
I dunno about this name. `is_sheaf_iff_is_sheaf_forget`?

Reply:
That's much nicer, thanks!